CS3219-SE-Principles-and-Patterns/cs3219-project-ay2122-2122-s1-g32
# Add Room Service and Protobuf

Completed:
- protobuf file and compile protobuf script for JavaScript and Go
- HTTP POST: /room/create, return `roomId`
- Websocket: JoinRoomRequest & JoinRoomResponse
   - check if user belongs to the room
   - return with `interviewerId` and `question`

Todo:
- [x] Websocket: CompleteQuestionRequest & CompleteQuestionResponse
- [x] HTTP POST: /history/meeting 
- [x] Websocket: SubmitRatingRequest & SubmitRatingResponse
- [x] HTTP POST: /history/rating


Test plan:
- install protobuf-compiler from [here](https://developers.google.com/protocol-buffers/docs/downloads) (download first one from release page, protobuf-all)
- run `.compile-proto-messages.sh` at project root directory
  - corresponding files should be generated in various service subdirectories based on the specification in the script
  - `yarn add google-protobuf` for JavaScript project    
  - if it is unable to compile for go, you may need to run `go mod vendor` in `./room`, but that shldnt affected the generated files for other languages

Sample Code Snippet for using generated protobuf types in JavaScript:

```javascript
const proto = require('./code2gather_pb');
let message = new proto.ClientRequest();
let joinRoomRequest = new proto.JoinRoomRequest();
joinRoomRequest.setRoomId('roomId');
message.setJoinRoomRequest(joinRoomRequest);
console.log(message.toString());
console.log(message.serializeBinary());
ws.send(message.serializeBinary(), { binary: true });


let message2 = proto.ClientRequest.deserializeBinary(message.serializeBinary());
console.log(message2.toObject());
console.log(proto.ErrorCode.NO_ERROR);
```
